### PR TITLE
Loading network screen

### LINKED ISSUE #3776: Improve test network loading state indicator
Bounty: when the extension is trying to connect to a Custom RPC network, we should

(1) update the UI to include the network name next to the small spinner on the top bar of the extension. Current behavior only shows a static spinner graphic, and most users don't know they can click it to escape.

(2) time out the connection attempt if it lasts longer than 15 seconds without connecting successfully. the extension should stop trying to connect and display the screen shown below. The "try again" button should resume attempting to connect to the same network, the "Switch networks" button should pop open the network menu (top-right).

current production:
screen shot 2018-10-12 at 12 04 33 pm

**design spec:**
(1) top-right network pill is updated, and UI shows the name of network
screen shot 2018-10-12 at 12 39 42 pm

(2) after 15s, modal over whatever previous screen the user was on. network should revert back to whatever the user was previously on if possible.
image

---

When a user clicks localhost 8545 without anything running on that port, they land on a spinner and think they're frozen. They can actually escape by clicking the small static spinner icon next to the top-right account menu, but few people realize this.

User feedback video showing the problem https://www.youtube.com/watch?v=Veeel3nUtbc

### LINKED ISSUE #5609: No evident UI to change network / login when perpetually "Connecting to Unknown Network"
**Describe the bug**
When MetaMask can't connect to the last network it was connected to, it has an overlay that blocks UI functionality that would allow a user to gracefully recover from the situation. 

**To Reproduce**
1. Install/log into MetaMask and connect to a custom network, for example local Ganache (http://127.0.0.1:7545). 
2. Do some transactions etc. to verify that everything is working fine as expected.
3. Log out of MetaMask, maybe close the browser.
4. Change to a state where that network you were connecting to is no longer accessible (e.g. close Ganache). 
5. Restart the browser.  Click the MetaMask button to log in.  You will see this: 
metamask forever connecting cannot change
There is no clearly reachable UI to log in or to change networks.  
There are hidden workarounds, but no clear UI to make them easy to identify. 
6. If possible, remember which network you were last connected to (the network selector UI that could remind you of which one it is is gone, so you'll need to rely on a good memory) and the requirements for accessing it, and then return to a state where you can connect to the network you were last connected to (in this case, restart Ganache).  _Note: It is not always possible to meet the requirements for accessing that previous network, e.g. if the previous URL was on a server that's now down or on an access-limited network that you are no longer connected to._
7. Again click the Metamask icon to attempt to log in. 
8. The results are the same as above; Metamask is stuck in this state without being able to determine that it can now connect.
9. Close and restart the browser.
10. Again click the Metamask icon to attempt to log in.  This time, you'll see: 
metamask welcome back
11. Log in and proceed.  

SECRET UI WORKAROUNDS IN STEP 5:
5a: If you start typing the password immediately after the click to open the window, and use the Enter key after a correct password, it will log in.  If you click in the password field first, intending to be sure the password field has input focus (generally a best practice for password entry), this will actually take focus away, because the overlay intercepted the click and took focus.  

5b: If you click the spinner in the upper right, you do get the network select list, but like 5a that's not at all clear UI.  However, clicking Custom RPC does nothing, other than dismiss the list.  You have to click Main Ethereum Network and wait for a connection to get the overlay to disappear, then log in, and THEN click the network selector for the desired network. 

If you are offline (e.g. on a plane or out somewhere without wifi), you won't be able to connect to the mainnet or any of the public testnets for workaround 5b.  If you're intending to connect to localhost (or another address on a semi-isolated network you're now connected to), but that's not a previously created option in the network selector list, too bad.  You won't be able to connect to any of the networks in the list and you can't access the Custom RPC UI to add it in.  You have to go through the 5a workaround and be sure not to click.  

**Other routes to reproduce**
12a. After logging in, open the network selector and choose one that you can't actually connect to at present (e.g. local Ganache, when Ganache is not running).  
12b. Alternatively to 12a, after logging in, open the network selector and choose CustomRPC, then enter an RPC URL where there is no responsive Ethereum node, and click Save.  
13. You will see the "Connecting to Unknown Network" overlay forever as before, again with the network selector UI gone.  
Screenshot for route 12a:
metamask connecting forever after login.  
Screenshot for route 12b: 
metamask connecting forever after login customrpc    
Assuming that users sometimes make errors when entering an RPC (e.g. choosing the wrong one of {`http`, `https`}, mistyping an IP or port number, using an incorrect separator, omitting the port number, etc., 12b is probably not that uncommon, and it should not leave the UI in an unusable state with no clear path to recovery.  
14. Closing the MetaMask popup and reopening it doesn't help.
15. You can click the identicon circle in the upper right and Log Out, but then you're back at step 5 above.  
16. You can also close your browser and restart it, and then you're back at step 5 above.  

**Expected behavior**
A. There is no "Connecting to ... Network" overlay on the login page.  
B. When the overlay is present (which per A should be only after login), the network selector is still visible and not replaced by a spinner.
C. The Custom RPC interface is accessible even when MetaMask cannot connect to whatever network it's trying to connect to, so that the user can switch to a network they _can_ connect to.  This is consistent with @danfinlay's [comment](https://github.com/MetaMask/metamask-extension/issues/1983#issuecomment-325442659) on issue #1983.  Note: Being able to access the CustomRPC UI while logged out, regardless of "Connecting..." overlay, is issue #5606; the expected behavior **here** is access to that UI when the "Connecting..." overlay is in place (even if logged in), instead of the overlay removing access to UI elements that allow logging in or changing RPC.    

**Browser details (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome 70.0.3538.77 (Official Build) (64-bit) (cohort: 70_77_Win)
 - MetaMask Version: 4.16.0
 - New / Beta UI

### PR DESCRIPTION
**Pull request by danjm**
fixes https://github.com/MetaMask/metamask-extension/issues/3776
fixes #5609 

peek 2018-12-07 21-14

### CODE REVIEW

File: ui/app/components/loading-network-screen/loading-network-screen.component.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import Spinner from '../spinner'
+import Button from '../button'
+
+class LoadingNetworkScreen extends Component {
+  constructor (props) {
+    super(props)
+
+    this.state = {
```
Comment (whymarrh) [inline ui/app/components/loading-network-screen/loading-network-screen.component.js]: We can move this out of the ctor and onto the class as a property

File: ui/app/components/loading-network-screen/loading-network-screen.component.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import Spinner from '../spinner'
+import Button from '../button'
+
+class LoadingNetworkScreen extends Component {
```
Comment (whymarrh) [inline ui/app/components/loading-network-screen/loading-network-screen.component.js]: Can we use `PureComponent` here?

File: ui/app/components/loading-network-screen/loading-network-screen.component.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            ? this.renderErrorScreenContent()
+            : this.renderLoadingScreenContent()
+          }
+        </div>
+      </div>
+    )
+  }
+}
+
+LoadingNetworkScreen.propTypes = {
```
Comment (whymarrh) [inline ui/app/components/loading-network-screen/loading-network-screen.component.js]: Can we can move this into the class a property?

**Review by whymarrh — CHANGES_REQUESTED** [review]
A couple of small nitpicky things. Also should we write `LoadingNetworkError` as a class for consistency sake?

File: ui/app/components/loading-network-screen/loading-network-screen.component.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      <span>Oops! Something went wrong.</span>
+      <div className="loading-overlay__error-buttons">
+        <Button
+          type="default"
+          onClick={() => {
+            window.clearTimeout(this.cancelCallTimeout)
+            showNetworkDropdown()
+          }}
+        >
+          { 'Switch Networks' }
```
Comment (alextsg) [inline ui/app/components/loading-network-screen/loading-network-screen.component.js]: Should this and "Try Again" below be translated strings?

File: ui/app/components/modals/loading-network-error/loading-network-error.component.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  const { t } = context
+  const { hideModal } = props
+
+  return (
+    <Modal
+      onSubmit={() => hideModal()}
+      submitText={t('tryAgain')}
+    >
+      <ModalContent
+        description={'Oops! Something went wrong.'}
```
Comment (alextsg) [inline ui/app/components/modals/loading-network-error/loading-network-error.component.js]: Should this be translated text?

File: ui/app/components/transaction-list/transaction-list.component.js [hunk 6]
```diff
@@ -40,7 +40,7 @@ export default class TransactionList extends PureComponent {
   shouldShowRetry = transaction => {
     const { transactionToRetry } = this.props
     const { id, submittedTime } = transaction
-    return id === transactionToRetry.id && Date.now() - submittedTime > 30000
+    return id === transactionToRetry.id && Date.now() - submittedTime > 3000
```
Comment (alextsg) [inline ui/app/components/transaction-list/transaction-list.component.js]: Should this change be committed? Or was the time reduced to 3 seconds for testing purposes?

File: ui/app/selectors.js [hunk 7]
```diff
@@ -36,10 +36,18 @@ const selectors = {
...
+  getNetworkIdentifier,
 }
 
 module.exports = selectors
 
+function getNetworkIdentifier (state) {
+  const { metamask: { provider: { type, nickname, rpcTarget } } } = state
+
+  return nickname || rpcTarget || type
+
```
Comment (alextsg) [inline ui/app/selectors.js]: nit: remove newline

**Review by alextsg — CHANGES_REQUESTED** [review]

**Comment by danjm** [comment]
@alextsg @whymarrh Addressed your comments in latest commit

**Comment by bdresser** [comment]
seeing a little funky text display `[SWITCHNETWORKS]`, maybe due to the translations? 

screen shot 2018-12-11 at 10 53 14 am

File: app/_locales/en/messages.json [hunk 8]
```diff
@@ -1215,6 +1221,8 @@
   },
   "speedUpTransaction": {
     "message": "Speed up this transaction"
+  "switchNetworks": {
```
Comment (whymarrh) [inline app/_locales/en/messages.json]: This looks like a formatting error?

Comment (whymarrh) [reply app/_locales/en/messages.json]: I think there should be a closing brace on the line above

**Review by whymarrh — CHANGES_REQUESTED** [review]
One small issue (I think)

**Review by whymarrh — APPROVED** [review]

**Review by alextsg — APPROVED** [review]

**Pull request merged**